### ISSUE TITLE
RFC: Always create `RouteSpec` string representation on construction

### ISSUE THREAD

**Issue by kyrias** [jbr/routefinder#7]
I noticed that the `Display` implementation for `RouteSpec` always builds up the string even if the `RouteSpec` has a `source` string specified and was about to send a PR that writes out the existing string when it's available.  But then I realized that it would probably make sense to make the `source` field no longer `Option`al (and probably rename it from `source` at the same time), and have the `From<Vec<Segment>>` implementation create the string representation.

This would mean that the string representation would only ever have to be constructed once for every `RouteSpec` instance rather than potentially every time you try to turn it into a string.

I thought of this because I wanted to make tide expose the matched route on every request, which currently means re-formatting the same string on every request, even though the same `RouteSpec` is used.

**Comment by jbr** [comment]
In the case of tide, source will always be Some, so it can be unwrapped and used for this purpose. Would that work? Alternatively, what about cloning the entire routespec and putting that in request extensions?

**Comment by kyrias** [comment]
> In the case of tide, source will always be Some, so it can be unwrapped and used for this purpose. Would that work?

Yes, but what about uses where it's _not_ created from a string?  It seems unnecessary to create the string representation every time it's asked for rather than once when the `RouteSpec` is created.  The second sentence of the first paragraph was the thing I was asking for comments on I guess.

> Alternatively, what about cloning the entire routespec and putting that in request extensions?

That's not really relevant to what I'm talking about as it would just push my question one step down the line.

**Comment by jbr** [comment]
> Yes, but what about uses where it's not created from a string? 

Is that possible in tide?

**Comment by kyrias** [comment]
Not currently, but tide isn't the only thing using this crate, and it feels to me like it would be more consistent to only build it once in every case.  But it doesn't matter too much to me, which is why this is an RFC.  :)

It would be neat if we could just do it on-demand in the implementation of `Display`, but that would require switching to some form of internal mutability like `ArcSwap`.
